state/git/impl.go

```diff
@@ -0,0 +1,193 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/kris-nova/kubicorn/apis/cluster"
+	"github.com/kris-nova/kubicorn/cutil/logger"
+	"github.com/kris-nova/kubicorn/state"
+
+	g "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+)
+
+type JSONGitStoreOptions struct {
+	BasePath     string
+	ClusterName  string
+	CommitConfig *JSONGitCommitConfig
+}
+
+type JSONGitCommitConfig struct {
+	Name   string
+	Email  string
+	Remote string
+}
+
+// JSONGitStore exists to save the cluster state as a git change.
+type JSONGitStore struct {
+	options      *JSONGitStoreOptions
+	ClusterName  string
+	BasePath     string
+	AbsolutePath string
+}
+
+func NewJSONGitStore(o *JSONGitStoreOptions) *JSONGitStore {
+	return &JSONGitStore{
+		options:      o,
+		ClusterName:  o.ClusterName,
+		BasePath:     o.BasePath,
+		AbsolutePath: filepath.Join(o.BasePath, o.ClusterName),
+	}
+}
+
+func (git *JSONGitStore) Exists() bool {
+	if _, err := os.Stat(git.AbsolutePath); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func (git *JSONGitStore) Write(relativePath string, data []byte) error {
+	fqn := filepath.Join(git.AbsolutePath, relativePath)
+	err := os.MkdirAll(path.Dir(fqn), 0700)
+	if err != nil {
+		return err
+	}
+	fo, err := os.Create(fqn)
+	if err != nil {
+		return err
+	}
+	defer fo.Close()
+
+	_, err = io.Copy(fo, strings.NewReader(string(data)))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (git *JSONGitStore) Read(relativePath string) ([]byte, error) {
+	fqn := filepath.Join(git.AbsolutePath, relativePath)
+	bytes, err := ioutil.ReadFile(fqn)
+
+	//if there's nothing check the remote repo
+	if err != nil {
+		// Create a new repository
+		r, _ := g.NewFilesystemRepository(fqn)
+		_, err = r.CreateRemote(&config.RemoteConfig{
+			Name: git.ClusterName,
+			URL:  git.options.CommitConfig.Remote,
+		})
+
+		// Pull using the create repository
+		err = r.Pull(&g.PullOptions{
+			RemoteName: git.ClusterName,
+		})
+		bytes, err = ioutil.ReadFile(fqn)
+		if err != nil {
+			return []byte(""), err
+		}
+	}
+
+	return bytes, nil
+}
+
+func (git *JSONGitStore) ReadStore() ([]byte, error) {
+	return git.Read(state.ClusterJSONFile)
+}
+
+//Performs a git 'commit' and 'push' of the current cluster changes.
+func (git *JSONGitStore) Commit(c *cluster.Cluster) error {
+	if c == nil {
+		return fmt.Errorf("Nil cluster spec")
+	}
+	bytes, err := json.Marshal(c)
+	if err != nil {
+		return err
+	}
+
+	//writes latest changes to git repo.
+	git.Write(state.ClusterJSONFile, bytes)
+
+	//commits the changes
+	r, err := g.NewFilesystemRepository(state.ClusterJSONFile)
+	if err != nil {
+		return err
+	}
+
+	// Add a new remote, with the default fetch refspec
+	_, err = r.CreateRemote(&config.RemoteConfig{
+		Name: git.ClusterName,
+		URL:  git.options.CommitConfig.Remote,
+	})
+	_, err = r.Commits()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (git *JSONGitStore) Rename(existingRelativePath, newRelativePath string) error {
```

Comment:
**Not required**

We could rename the state in the remote as well. Currently, this will rename state on local machine but will not push it to the remote.

Reply:
Renaming the remote state could make this more complex, which makes things more complex over time. It's fine as it is.